src/uct/tcp/tcp_listener.c

```diff
@@ -171,11 +172,29 @@ ucs_status_t uct_tcp_listener_reject(uct_listener_h listener,
               cep, cep->fd, cep->state,
               uct_tcp_sockcm_cm_ep_peer_addr_str(cep, peer_str, UCS_SOCKADDR_STRING_LEN));
 
-    status = UCS_OK;
+    hdr = (uct_tcp_sockcm_priv_data_hdr_t*)cep->comm_ctx.buf;
+
+    hdr->length          = 0;   /* sending only the header in the reject message */
+    hdr->status          = (uint8_t)UCS_ERR_REJECTED;
+    cep->comm_ctx.length = sizeof(*hdr);
+
+    /* the server needs to send a reject message to the client */
+    ucs_async_modify_handler(cep->fd, UCS_EVENT_SET_EVWRITE | UCS_EVENT_SET_EVREAD);
+
+    cep->state |= UCT_TCP_SOCKCM_EP_SERVER_REJECT_CALLED;
+    status = uct_tcp_sockcm_ep_progress_send(cep);
+
 out:
+    if (status != UCS_OK) {
+        if (status == UCS_ERR_REJECTED) {
+            ucs_assert(cep->state & UCT_TCP_SOCKCM_EP_SERVER_REJECT_SENT);
+            status = UCS_OK;
+        }
+
+        UCS_CLASS_DELETE(uct_tcp_sockcm_ep_t, cep);
```

Comment:
seems ugly, need to use standard error flow with goto label for each cleanup phase